File: ios/Sources/MritMesh/DSL/Mrit.swift
import Foundation

/// Mrit — the high-level, app-facing entry point to the MRIT mesh (Phase 5).
///
/// `MeshNode` is the Layer 2/3 protocol engine: packets, routing, encryption,
/// store-and-forward. `Mrit` wraps it in a small declarative configuration API
/// so application code never touches a `MeshPacket` directly.
///
/// Mirrors the Android `Mrit` DSL in
/// `app/src/main/java/com/mrit/mesh/dsl/Mrit.kt` — the two are intentionally
/// symmetric so a shared app design doc reads the same on both platforms.
///
/// Example:
/// ```swift
/// let mesh = Mrit { config in
///     config.onMessage { msg in print("From \(msg.from.shortId): \(msg.text)") }
///     config.onFile    { file in save(file.fileName, file.data) }
///     config.onPeers   { peers in updatePeerList(peers) }
/// }
///
/// mesh.send(to: peerId, text: "Hello mesh!")
/// mesh.sendFile(to: peerId, name: "map.png", data: bytes)
/// mesh.sos("Need help — twisted ankle, 2km north of trailhead")
///
/// // ... later
/// mesh.stop()
/// ```
public final class Mrit {

    /// The underlying protocol engine. Most apps never need to touch this directly.
    public let node: MeshNode

    /// Our own identity on the mesh — share this with others so they can message us.
    public var ourId: MeshID { node.ourId }

    /// Create, configure, and start a mesh node in one call.
    ///
    /// - Parameter configure: closure registering message/file/peer handlers,
    ///   evaluated once before the node starts.
    public init(configure: (MritConfig) -> Void = { _ in }) {
        let config = MritConfig()
        configure(config)

        let node = MeshNode()
        node.onMessageReceived = { msg in config.messageHandlers.forEach { $0(msg) } }
        node.onFileReceived    = { name, data in
            let file = ReceivedFile(fileName: name, data: data)
            config.fileHandlers.forEach { $0(file) }
        }
        node.onPeersChanged = { peers in config.peerHandlers.forEach { $0(peers) } }

        self.node = node
        node.start()
    }

    // ── Send API ───────────────────────────────────────────────────────────────

    /// Send an end-to-end encrypted text message to `dest`. Routes multi-hop automatically.
    public func send(to dest: MeshID, text: String) {
        node.sendMessage(to: dest, text: text)
    }

    /// Send a file of any size to `dest` — automatically chunked (32KB) and encrypted.
    public func sendFile(to dest: MeshID, name: String, data: Data) {
        node.sendFile(to: dest, name: name, data: data)
    }

    /// Broadcast an unencrypted emergency message to every reachable node.
    public func sos(_ message: String) {
        node.sendSOS(message)
    }

    // ── Lifecycle ──────────────────────────────────────────────────────────────

    /// Stop the mesh node and release all transport resources.
    public func stop() {
        node.stop()
    }
}

/// Configuration block for `Mrit.init`.
///
/// Register one or more handlers per event type — every registered handler is
/// called for every event of that type.
public final class MritConfig {
    fileprivate var messageHandlers: [(MeshNode.IncomingMessage) -> Void] = []
    fileprivate var fileHandlers:    [(ReceivedFile) -> Void] = []
    fileprivate var peerHandlers:    [([MeshNode.PeerInfo]) -> Void] = []

    public init() {}

    /// Called for every text message (including SOS broadcasts) addressed to or seen by us.
    public func onMessage(_ handler: @escaping (MeshNode.IncomingMessage) -> Void) {
        messageHandlers.append(handler)
    }

    /// Called once per file transfer, after all chunks have arrived and been reassembled.
    public func onFile(_ handler: @escaping (ReceivedFile) -> Void) {
        fileHandlers.append(handler)
    }

    /// Called whenever the set of reachable peers changes (peer joins, leaves, or times out).
    public func onPeers(_ handler: @escaping ([MeshNode.PeerInfo]) -> Void) {
        peerHandlers.append(handler)
    }
}

/// A fully reassembled file received from a peer.
/// Mirrors Android's `MeshNode.ReceivedFile`.
public struct ReceivedFile {
    public let fileName: String
    public let data: Data
}

File: app/src/main/java/com/mrit/mesh/dsl/Mrit.kt
package com.mrit.mesh.dsl

import android.content.Context
import com.mrit.mesh.core.MeshID
import com.mrit.mesh.core.PeerInfo
import com.mrit.mesh.mesh.MeshNode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

/**
 * Mrit — the high-level, app-facing entry point to the MRIT mesh (Phase 5).
 *
 * [MeshNode] is the Layer 2/3 protocol engine: packets, routing, encryption,
 * store-and-forward. [Mrit] wraps it in a small declarative configuration DSL
 * so application code never touches a [com.mrit.mesh.core.MeshPacket] directly.
 *
 * Mirrors the iOS `Mrit` API in `ios/Sources/MritMesh/DSL/Mrit.swift` —
 * the two are intentionally symmetric so a shared app design doc reads the
 * same on both platforms.
 *
 * Example:
 * ```kotlin
 * val mesh = Mrit.start(applicationContext) {
 *     onMessage { msg  -> log("From ${msg.from.shortId()}: ${msg.text}") }
 *     onFile    { file -> saveToDisk(file.fileName, file.data) }
 *     onPeers   { peers -> updatePeerChips(peers) }
 * }
 *
 * mesh.send(peerId, "Hello mesh!")
 * mesh.sendFile(peerId, "map.png", bytes)
 * mesh.sos("Need help — twisted ankle, 2km north of trailhead")
 *
 * // ... later, e.g. in onDestroy()
 * mesh.stop()
 * ```
 */
class Mrit private constructor(val node: MeshNode) {

    private val scope = CoroutineScope(Dispatchers.Main.immediate + SupervisorJob())

    companion object {
        /**
         * Create, configure, and start a mesh node in one call.
         *
         * @param context   Android context (application context recommended —
         *                   the underlying [MeshNode] outlives any single Activity)
         * @param configure DSL block registering message/file/peer handlers,
         *                   evaluated once before the node starts
         * @return a started [Mrit] instance ready to send messages
         */
        fun start(context: Context, configure: MritConfig.() -> Unit = {}): Mrit {
            val config = MritConfig().apply(configure)
            val node = MeshNode(context.applicationContext)
            val mrit = Mrit(node)

            mrit.scope.launch {
                node.incomingMessages.collect { msg -> config.messageHandlers.forEach { it(msg) } }
            }
            mrit.scope.launch {
                node.incomingFiles.collect { file -> config.fileHandlers.forEach { it(file) } }
            }
            mrit.scope.launch {
                node.peers.collect { peers -> config.peerHandlers.forEach { it(peers) } }
            }

            node.start()
            return mrit
        }
    }

    // ── Identity ───────────────────────────────────────────────────────────────

    /** Our own identity on the mesh — share this with others so they can message us. */
    val ourId: MeshID get() = node.ourId

    /** Currently reachable peers — updates live as devices come in/out of range. */
    val peers: StateFlow<List<PeerInfo>> get() = node.peers

    // ── Send API ───────────────────────────────────────────────────────────────

    /** Send an end-to-end encrypted text message to [to]. Routes multi-hop automatically. */
    fun send(to: MeshID, text: String) = node.sendMessage(to, text)

    /** Send a file of any size to [to] — automatically chunked (32KB) and encrypted. */
    fun sendFile(to: MeshID, fileName: String, data: ByteArray) = node.sendFile(to, fileName, data)

    /** Broadcast an unencrypted emergency message to every reachable node. */
    fun sos(message: String) = node.sendSOS(message)

    // ── Lifecycle ──────────────────────────────────────────────────────────────

    /** Stop the mesh node and release all transport resources (WiFi Direct, BLE). */
    fun stop() {
        scope.cancel()
        node.stop()
    }
}

/**
 * DSL configuration block for [Mrit.start].
 *
 * Register one or more handlers per event type — every registered handler is
 * called for every event of that type. Handlers run on the main thread.
 */
class MritConfig {
    internal val messageHandlers = mutableListOf<(MeshNode.IncomingMessage) -> Unit>()
    internal val fileHandlers    = mutableListOf<(MeshNode.ReceivedFile) -> Unit>()
    internal val peerHandlers    = mutableListOf<(List<PeerInfo>) -> Unit>()

    /** Called for every text message (including SOS broadcasts) addressed to or seen by us. */
    fun onMessage(handler: (MeshNode.IncomingMessage) -> Unit) {
        messageHandlers += handler
    }

    /** Called once per file transfer, after all chunks have arrived and been reassembled. */
    fun onFile(handler: (MeshNode.ReceivedFile) -> Unit) {
        fileHandlers += handler
    }

    /** Called whenever the set of reachable peers changes (peer joins, leaves, or times out). */
    fun onPeers(handler: (List<PeerInfo>) -> Unit) {
        peerHandlers += handler
    }
}
